astropy/coordinates/transformations.py

```diff
@@ -746,8 +749,157 @@ def __call__(self, fromcoord, toframe):
         if not isinstance(res, self.tosys):
             raise TypeError('the transformation function yielded {0} but '
                 'should have been of type {1}'.format(res, self.tosys))
+        if (getattr(fromcoord.data, 'differentials', None) and not
+            getattr(res.data, 'differentials', None)):
+            warn("Applied a FunctionTransform to a coordinate frame with "
+                 "differentials, but the FunctionTransform does not handle "
+                 "differentials, so they have been dropped.", AstropyWarning)
         return res
 
+class FunctionTransformWithFiniteDifference(FunctionTransform):
+    """
+    A coordinate transformation that works like a `FunctionTransform`, but
+    computes velocity shifts based on the finite-difference relative to one of
+    the frame attributes.  Note that the transform function should *not* change
+    the differential at all in this case, as any differentials will be
+    overridden.
+
+    Parameters
+    ----------
+    finite_difference_frameattr_name : str or None
+        The name of the frame attribute on the frames to use for the finite
+        difference.  Both the to and the from frame will be checked for this
+        attribute, but only one needs to have it. If None, only the
+        re-orientation of the existing differential will be computed.
+    finite_difference_dt : Quantity or callable
+        If a quantity, this is the size of the differential used to do the
+        finite difference.  If a callable, should accept
+        ``(fromcoord, toframe)`` and return the ``dt`` value.
+    symmetric_finite_difference : bool
+        If True, the finite difference is computed as
+        :math:`\frac{x(t + \Delta t / 2) - x(t + \Delta t / 2)}{\Delta t}`, or
+        if False, :math:`\frac{x(t + \Delta t) - x(t)}{\Delta t}`.  The latter
+        case has slightly better performance (and more stable finite difference
+        behavior).
+
+    All other paramters are identical to the initializer for
+    `FunctionTransform`.
+
+    """
+    def __init__(self, func, fromsys, tosys, priority=1, register_graph=None,
+                 finite_difference_frameattr_name='obstime',
+                 finite_difference_dt=1*u.second,
+                 symmetric_finite_difference=True):
+        super(FunctionTransformWithFiniteDifference, self).__init__(func,
+              fromsys, tosys, priority, register_graph)
+        self.finite_difference_frameattr_name = finite_difference_frameattr_name
+        self.finite_difference_dt = finite_difference_dt
+        self.symmetric_finite_difference = symmetric_finite_difference
+
+    @property
+    def finite_difference_frameattr_name(self):
+        return self._finite_difference_frameattr_name
+    @finite_difference_frameattr_name.setter
```

Comment:
actually I favor this construction because it's clear it's all part of the same property...